proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,107 @@
+mistake in deciding the latest optimistically confirmed slot, it could mean
+rollback of user transactions after they have been confirmed, which is not
+acceptable.
+
+We aim to automate the finding of oc slot and snapshot generation, so that
+we can lower the possibility of human mistakes in the cluster restart process.
+
+## Alternatives Considered
+
+See [Handling Common Solana Outages](https://docs.google.com/document/d/1RkNAyz-5aKvv5FF44b8SoKifChKB705y5SdcEoqMPIc/edit#bookmark=id.jtrjbe6g4mk3) for details.
+
+There are many proposals about automatically detecting that the cluster is
+in an outage so validators should enter a recovery process automatically.
+
+While getting human out of the loop greatly improves recovery speed,
+automaticlly restarting the whole cluster still seems risky. Because if
+the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+## New Terminology
+
+* SilentRepairMode - when validators restart in this new mode, they will
+talk with each other to find the highest oc slot and repair all necessary
+blocks. To improve speed and guarantee simplicity, Turbine, vote, and new
+block generation are paused in this mode.
+
+## Detailed Design
+
+A new command line arg --RepairAndRestart <optional_slots_to_send_on_last_voted_fork> is
+added. When the cluster is in need of a restart, we assume at least 80% will
+restart with this arg. Any validators restarted with this arg does not participate
+in the normal Turbine protocol, update its vote, or generate new blocks until all
+of the following steps are complted.
+
+### Gossip last vote before the restart and ancestors on that fork
+Send Gossip message LastVotedForkSlots to everyone, it contains the last voted slot on
+its tower and the ancestor slots on the last voted fork and is sent in a bitmap like
+the EpicSlots data structure. The number of ancestor slots sent is determined by
+<optional_slots_to_send_on_last_voted_fork>, by default this number is 2000.
+
+### Aggregate, repair, and replay the slots in LastVotedForkSlots
+Aggregate the slots in received LastVotedForkSlots messages, whenever some slot has
+more than 47% stake and it's missing locally, start the repair process for this slot.
+47% is chosen because we don't want to miss any optimistically confirmed slot, those
+slots have at least 67% votes, and we assume that there can be 20% validators not
+participating in restarts, so we need to repair all slots with at least
+67% - 20% = 47% votes.
+
+### Gossip current heaviest fork
+After receiving LastVotedForkSlots from 80% of the validators and reparing all slots
+with more than 47% votes, count the heaviest fork and Gossip Heaviest(X, Hash(X)) out,
+where X is the tip of the heaviest fork.
+
+### Generate local snapshot on the heaviest slot
+Generate a local snapshot on the heaviest slot.
```

Comment:
This is done because in case people can't agree on the restart slots, people on the correct heaviest slot can restart in the old way faster using the local snapshot. People n the wrong heaviest slot can potentially download the snapshot from others.

Reply:
hmm, we can only generate snapshots of a root currently

Reply:
Why is that? Is this something that can be changed?

Reply:
> Why is that? 

Limitation of the current snapshot format

> Is this something that can be changed?

Yep, it's just software. Currently a snapshot contains a single bank and is all that's populated into `BankForks` at startup, so it's effectively considered a "root" - in that that bank will never be rolled back at runtime.

Supporting multiple banks in a snapshot certainly is desirable longer term, but would be a significant detour if support for such is deemed to be mandatory to implement this proposal